### PR TITLE
feat: support git hosted package dependencies when versioning

### PR DESCRIPTION
**Pull request by fstof**
## Description
This change implements the feature request #159 

When packages are hosted on git (ie, not published). When we call `melos version` the package's `pubspec.yaml` git ref tags should be updated to the latest git tag created.

## Type of Change

- [x] ✨ `feat` -- New feature (non-breaking change which adds functionality)
- [ ] 🛠️ `fix` -- Bug fix (non-breaking change which fixes an issue)
- [ ] ❌ `!` -- Breaking change (fix or feature that would cause existing functionality to change)
- [ ] 🧹 `refactor` -- Code refactor
- [ ] ✅ `ci` -- Build configuration change
- [x] 📝 `docs` -- Documentation
- [ ] 🗑️ `chore` -- Chore

### CODE REVIEW

File: packages/melos/lib/src/commands/version.dart [hunk 1]
```diff
@@ -439,10 +441,23 @@ Hint: try running "melos version --all" to include private packages.
...
-        dependencyVersionReplaceRegex(dependencyName), (Match match) {
-      return '${match.group(1)}$dependencyVersion';
-    });
+    String updatedContents;
+
+    final gitReference =
+        package.pubSpec.dependencies[dependencyName] is GitReference ||
+            package.pubSpec.devDependencies[dependencyName] is GitReference;
+
+    if (gitReference) {
```
Comment (Salakar) [inline packages/melos/lib/src/commands/version.dart:450]: We should probably only do this if `--git-tag-version` is true right?

**Comment by Salakar** [comment]
Thanks for this, I think if we want to support this we should put it behind an melos.yaml option for the version command, e.g. like these: https://github.com/invertase/melos/blob/main/melos.yaml#L7-L8 - since it's very specific behaviour (and also would break current behaviour). Thoughts?

File: packages/melos/lib/src/commands/version.dart (hunk 1 above)
Comment (fstof) [reply packages/melos/lib/src/commands/version.dart:450]: That makes sense yes.
I will check how we do that and update

Comment (fstof) [reply packages/melos/lib/src/commands/version.dart:450]: Okay @Salakar 
I double checked this flag. (The documentation needs updating as it is just a copy/paste from `--no-changelog`) 

However. It seems like this option still updates the versions in `pubspec.yaml` only skips the git commit (which obviously also does not tag).

If this is indeed the intended behaviour (skipping the commit) then I think we should still do the update, but do introduce the cofig in `melos.yaml`

**Comment by fstof** [comment]
> Thanks for this, I think if we want to support this we should have it turned off by default and put it behind an `melos.yaml` option for the version command, e.g. like these: https://github.com/invertase/melos/blob/main/melos.yaml#L7-L8 - since it's very specific behaviour (and also would break current behaviour). Thoughts?

Hi @Salakar
So I implemented both suggestions, (the `melos.yaml` config and `--git-tag-version` check) you can check if you hare happy with that.

But I am of the opinion that `--git-tag-version` is not required as i mentioned in https://github.com/invertase/melos/pull/256#discussion_r820081584

File: docs/guides/automated-releases.mdx [hunk 2]
```diff
@@ -53,3 +53,33 @@ Once satisfied with your pending releases, release them to [pub.dev](https://pub
 ```dart
 melos publish --no-dry-run
 ```
+
+## git hosted packages
```
Comment (Salakar) [inline docs/guides/automated-releases.mdx:57]:
```suggestion
## Git hosted packages
```

File: docs/guides/automated-releases.mdx [hunk 3]
```diff
@@ -53,3 +53,33 @@ Once satisfied with your pending releases, release them to [pub.dev](https://pub
...
+```yaml
+dependencies:
+  internal_dep:
+    git:
+      url: git@github.com:org/repo.git
+      path: packages/internal_dep
+      ref: internal_dep-v0.0.1
+
+# to be updated to
+
```
Comment (Salakar) [inline docs/guides/automated-releases.mdx:78]:
```suggestion

# will be updated to:

```

**Comment by fstof** [comment]
> > But I am of the opinion that --git-tag-version is not required as i mentioned in [#256 (comment)](https://github.com/invertase/melos/pull/256#discussion_r820081584)
> 
> Let's remove this check then since it's now going to be behind melos.yaml option anyway

removed the `--git-tag-version` check

**Comment by fstof** [comment]
@Salakar 

Is there anything specific still holding this back?

**Review by Salakar — APPROVED** [review]

**Pull request merged**